src/Pos/Client/Txp/History.hs

```diff
@@ -78,84 +80,78 @@ data TxHistoryAnswer = TxHistoryAnswer
 ----------------------------------------------------------------------
 
 -- | Check if given 'Address' is one of the receivers of 'Tx'
-hasReceiver :: Tx -> Address -> Bool
-hasReceiver UnsafeTx {..} addr = any ((== addr) . txOutAddress) _txOutputs
+hasReceiver :: Tx -> Set Address -> Set Address
+hasReceiver UnsafeTx {..} addrs =
+    addrs `S.intersection` (S.fromList $ toList (txOutAddress <$> _txOutputs))
 
 -- | Given some 'Utxo', check if given 'Address' is one of the senders of 'Tx'
-hasSender :: MonadUtxoRead m => Tx -> Address -> m Bool
-hasSender UnsafeTx {..} addr = anyM hasCorrespondingOutput $ toList _txInputs
-  where hasCorrespondingOutput txIn =
-            fmap toBool $ ((== addr) . txOutAddress . toaOut) <<$>> utxoGet txIn
-        toBool Nothing  = False
-        toBool (Just b) = b
+hasSender :: MonadUtxoRead m => Tx -> Set Address -> m [Address]
+hasSender UnsafeTx {..} addrs =
+  catMaybes <$> mapM fetchWithCorrespondingOutput (toList _txInputs)
+  where fetchWithCorrespondingOutput txIn = runMaybeT $ do
+          inAddr <- MaybeT $ txOutAddress . toaOut <<$>> utxoGet txIn
+          guard (inAddr `S.member` addrs)
+          return inAddr
 
 -- | Datatype for returning info about tx history
 data TxHistoryEntry = THEntry
     { _thTxId       :: !TxId
     , _thTx         :: !Tx
     , _thIsOutput   :: !Bool
     , _thDifficulty :: !(Maybe ChainDifficulty)
+    , _thInputAddr  :: !Address
     } deriving (Show, Eq, Generic)
 
 makeLenses ''TxHistoryEntry
 
 -- | Type of monad used to deduce history
 type TxSelectorT m = UtxoStateT (MaybeT m)
 
--- | Select transactions related to given address. `Bool` indicates
+-- | Select transactions related to given addresses. `Bool` indicates
 -- whether the transaction is outgoing (i. e. is sent from given address)
```

Comment:
Well, not looking into the code, but in general:
> single transaction may only have single source address
this is false. As one transaction in cardano can have multiple inputs and multiple outputs. Are you referring to this? I can have address A and address B and I can create transaction within cardano which would spend from both A and B and send to C and D (so two inputs and two outputs in this case).

As of the discussion, I was just referring to comments/description of the function that seemed outdated with new implementation. I wasn't very carefully review some parts of the code as I don't have full picture of some parts of the code base (as this function specifically), nor about HD wallets so far. But I will review one more time after I re-read specs and implement client-api, so hope I will give more informed review.

Again, cardano can have many-to-many transactions. With HD-wallets we will sure have transactions with multiple source addresses. Imaging wallet/account that has 2 addresses:
```
addressA   10ada
addressB   5ada
```
and you would like to send 12 ada from this wallet to me. We will do this by creating one transactions that takes both `addressA` and `addressB` as inputs and send 12ada to my address. We can implement algorithm how to chose from which wallets to spend in different way: 10ada from addressA + 2ada from addressB = 12 ada, or 7ada from addressA + 5ada from addressB = 12 ada, ...